lib/Microservice.js

```diff
@@ -0,0 +1,139 @@
+const { promisify } = require('util');
+const loopback = require('loopback');
+const boot = require('loopback-boot');
+
+/**
+ * Basic Microservice class wrapping a loopback application.
+ *
+ * @type {module.Microservice}
+ */
+module.exports = class Microservice {
+
+    constructor(app, bootOptions = {}) {
+        const defaultAppDir = './app';
+        const defaultConfigDir = `${defaultAppDir}/config`;
+
+        this.app = app;
+        this.server = null;
+        this._bootOptions = Object.assign({
+            appRootDir: defaultAppDir,
+            appConfigRootDir: defaultConfigDir,
+            componentRootDir: defaultConfigDir,
+            dsRootDir: defaultConfigDir,
+            env: 'dev',
+            middlewareRootDir: defaultConfigDir,
+            modelsRootDir: defaultConfigDir,
+        }, bootOptions);
+    }
+
+    /**
+     * Stop the webserver of the microservice.
+     *
+     * @returns {Promise.<Microservice>}
+     */
+    async stop() {
+
+        if (!this.server || !this.server.listening) return Promise.reject(new Error('Microservice not running'));
+
+        const server = this.server;
+        // avoid race conditions
+        this.server = null;
+
+        return new Promise((resolve, reject) => {
+            // does not work with promisify
+            server.close((error) => {
+                if (error) {
+                    // if unable to stop the server, reassign it.
+                    this.server = server;
+                    return reject(error);
+                }
+                resolve(this);
+            });
+        });
+    }
+
+    /**
+     * Start the webserver of the microservice.
+     *
+     * @returns {Promise.<Microservice>}
+     */
+    async start() {
+
+        if (this.server && this.server.listening) return Promise.resolve(this);
+
+        return new Promise((resolve, reject) => {
+            // @note: this method might throw an exception, we should not catch it,
+            //        due to possible issues in the testing
+            this.server = this.app.listen(() => {
+
+                this.app.emit('started');
+
+                const baseUrl = this.app.get('url').replace(/\/$/, '');
+                const logger = this.app.get('logger');
+
+                logger.info('Web server listening at: %s', baseUrl);
+
+                if (this.app.get('loopback-component-explorer')) {
+                    const explorerPath = this.app.get('loopback-component-explorer').mountPath;
+                    logger.info('Browse your REST API at %s%s', baseUrl, explorerPath);
+                }
+
+                resolve(this);
+            });
+        });
+
+    }
+
+    /**
+     * Boots the application.
+     *
+     * More information on the options can be found here https://apidocs.strongloop.com/loopback-boot/.
+     *
+     * @todo: Do we need to protect the app from being booted multiple times?
```

Comment:
pleas dont use todos in pull requests see https://github.com/joinbox/guidelines/blob/master/styleguide/javascript.md 18.4